src/lib/operators/join_mpsm.cpp

```diff
@@ -0,0 +1,520 @@
+}
+
+std::shared_ptr<const Table> JoinMPSM::_on_execute() {
+  // Check column types
+  const auto& left_column_type = input_table_left()->column_data_type(_column_ids.first);
+  DebugAssert(left_column_type == input_table_right()->column_data_type(_column_ids.second),
+              "Left and right column types do not match. The mpsm join requires matching column types");
+
+  // Create implementation to compute the join result
+  _impl = make_unique_by_data_type<AbstractJoinOperatorImpl, JoinMPSMImpl>(
+      left_column_type, *this, _column_ids.first, _column_ids.second, _predicate_condition, _mode);
+
+  return _impl->_on_execute();
+}
+
+void JoinMPSM::_on_cleanup() { _impl.reset(); }
+
+std::shared_ptr<AbstractOperator> JoinMPSM::_on_recreate(
+    const std::vector<AllParameterVariant>& args, const std::shared_ptr<AbstractOperator>& recreated_input_left,
+    const std::shared_ptr<AbstractOperator>& recreated_input_right) const {
+  return std::make_shared<JoinMPSM>(recreated_input_left, recreated_input_right, _mode, _column_ids,
+                                         _predicate_condition);
+}
+
+const std::string JoinMPSM::name() const { return "Join MPSM"; }
+
+template <typename T>
+class JoinMPSM::JoinMPSMImpl : public AbstractJoinOperatorImpl {
+ public:
+  JoinMPSMImpl<T>(JoinMPSM& sort_merge_join, ColumnID left_column_id, ColumnID right_column_id,
+                  const PredicateCondition op, JoinMode mode)
+      : _mpsm_join{sort_merge_join},
+        _left_column_id{left_column_id},
+        _right_column_id{right_column_id},
+        _op{op},
+        _mode{mode} {
+    _cluster_count = _determine_number_of_clusters();
+    _output_pos_lists_left.resize(_cluster_count);
+    _output_pos_lists_right.resize(_cluster_count);
+    for (size_t cluster = 0; cluster < _cluster_count; ++cluster) {
+      _output_pos_lists_left[cluster].resize(1);
+      _output_pos_lists_right[cluster].resize(_cluster_count);
+    }
+  }
+
+ protected:
+  JoinMPSM& _mpsm_join;
+
+  // Contains the materialized sorted input tables
+  std::unique_ptr<MaterializedNUMAPartitionList<T>> _sorted_left_table;
+  std::unique_ptr<MaterializedNUMAPartitionList<T>> _sorted_right_table;
+
+  // Contains the null value row ids if a join column is an outer join column
+  std::unique_ptr<PosList> _null_rows_left;
+  std::unique_ptr<PosList> _null_rows_right;
+
+  const ColumnID _left_column_id;
+  const ColumnID _right_column_id;
+
+  const PredicateCondition _op;
+  const JoinMode _mode;
+
+  // the cluster count must be a power of two, i.e. 1, 2, 4, 8, 16, ...
+  ClusterID _cluster_count;
+
+  // Contains the output row ids for each cluster
+  std::vector<std::vector<std::shared_ptr<PosList>>> _output_pos_lists_left;
+  std::vector<std::vector<std::shared_ptr<PosList>>> _output_pos_lists_right;
+
+  /**
+   * The TablePosition is a utility struct that is used to define a specific position in a sorted input table.
```

Comment:
I have a concept of what a row and a column is, but not what a "specific position in a sorted table" is. Can you pin that down more precisely?